nemo_aligner/data/nlp/datasets.py

```diff
@@ -403,10 +432,153 @@ def __getitem__(self, idx):
             "rejected_labels": labels_reject_tokens,
             "chosen_reward": payload.get("chosen_reward", self.default_chosen_reward),
             "rejected_reward": payload.get("rejected_reward", self.default_rejected_reward),
+            "ignore_example": ignore_example,
         }
         return output
 
 
+class DPOPackedDataset(DPOModelDataset):
+    """A dataset class for DPO with sequence packing. Data is expected to be 
+    pre-tokenized and pre-packed using examples/nlp/data/dpo/prepare_packed_dpo_dataset.py.
+    """
+
+    def __init__(
+        self,
+        cfg,
+        tokenizer,
+        name,
+        data_prefix,
+        documents,
+        data,
+        seq_length,
+        seed,
+        drop_last=True,  # return_cu_seqlen: bool = True ## should always be true
+    ):
+
+        super().__init__(cfg, tokenizer, name, data_prefix, documents, data, seq_length, seed, drop_last)
+        self.data_prefix = data_prefix
+
+    def __getitem__(self, idx):
+        return self.data[idx]
+
+    def __len__(self):
+        return len(self.data)
+
+    def _ceil_to_nearest(self, n, m):
+        return (n + m - 1) // m * m
+
+    def _maybe_cast_to_list(self, x):
+        return [item.tolist() if isinstance(item, np.ndarray) else item for item in x]
+
+    def _collate_item(self, item, max_length, pad_id):
+        item = self._maybe_cast_to_list(item)
+        item = [x + [pad_id] * (max_length - len(x)) for x in item]
+        return item
+
+    ## reset_position_ids, reset_attention_mask and eod_mask_loss are unused but are needed to match the API of dpo_custom_collate
+    def collate_fn(
```

Comment:
could we maybe name this "global_collate_fn" so that it's not confused with the collate_fn nomenclature of torch datasets which is used within the multiprocessing context?